source/common/common/numeric.h

```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <type_traits>
+
+namespace Envoy {
+
+// gcd() defined here is a substitute for C++17's implementation of std::gcd() from <numeric>,
+// loosely adapted from http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4061.pdf.
+// NOTE: std::abs() defined in <cmath> and <cstdlib> does not support arbitrary integral types
+//       including, perhaps for obvious reasons, unsigned integers.
+
+template <typename T> constexpr auto abs(T n) -> std::enable_if_t<std::is_unsigned<T>::value, T> {
+  return n;
+}
+
+template <typename T>
+constexpr auto abs(T n)
+    -> std::enable_if_t<std::is_integral<T>::value and !std::is_unsigned<T>::value, T> {
+  return n < T(0) ? -n : n;
+}
+
+template <typename T>
+constexpr auto gcd(T m, T n) -> std::enable_if_t<std::is_integral<T>::value, T> {
```

Comment:
I'd add a `TODO` pointing out that this will all be removed in C++17 and call sites should be adapted then.

Also, since you're new to the code base, I'd point out that this extensive use of template type inference etc. is not normal in the code base. It's OK for a few language base utils, but best avoided (see style guide).

Reply:
Yeah, I'm sort of disappointed that I have to do this at all. There are a few directions I could go with this:

* The simplest thing that would meet my actual needs would be to implement `gcd()` only for `uint32_t` — no templates at all — maybe as a static member function of Ring.
* Something more along the lines of what I've done here, but without the type_traits stuff, would be just to say `template <typename T> constexpr T gcd(T m, T n)` and not worry too much about it.
* Find out how long it takes to add `gcd()` to Abseil :)

Opinions?

Reply:
It's fine as is, just sharing a PSA tip.